packages/rsocket-core/src/ClientServerMultiplexerDemultiplexer.ts

```diff
@@ -318,14 +360,99 @@ class ConnectionFrameHandler implements FrameHandler {
       // TODO: throw an exception and close connection
     }
   }
+
+  close(error?: Error) {
+    this.keepAliveHandler.close();
+    this.rsocket.close?.call(this.rsocket, error);
+  }
+}
+
+class KeepAliveHandler implements FrameHandler {
+  private keepAliveLastReceivedMillis = Date.now();
+  private activeTimeout: any;
+
+  constructor(
+    private readonly multiplexer: ClientServerInputMultiplexerDemultiplexer,
+    private readonly outbound: Outbound,
+    private keepAliveTimeoutDuration: number
+  ) {
+    this.activeTimeout = setTimeout(
+      this.timeoutCheck.bind(this),
+      keepAliveTimeoutDuration
+    );
+  }
+
+  handle(frame: KeepAliveFrame): void {
+    this.keepAliveLastReceivedMillis = Date.now();
+    if (Flags.hasRespond(frame.flags)) {
+      this.outbound.send({
+        type: FrameTypes.KEEPALIVE,
+        streamId: 0,
+        data: undefined,
```

Comment:
Sending `undefined` in the data may not conform to the protocol.

https://rsocket.io/about/protocol#keepalive-frame-0x03

> Reception of a KEEPALIVE frame with the (R)espond flag set MUST cause a client or server to send back a KEEPALIVE with the (R)espond flag NOT set. The data in the received KEEPALIVE MUST be echoed back in the generated KEEPALIVE.

**The data in the received KEEPALIVE MUST be echoed back in the generated KEEPALIVE.**